### ISSUE TITLE
Hopper FNA backward produces NaN in dQ over a partial last KV tile (FP16)

### ISSUE THREAD

**Issue by luowyang** [luowyang/NATTEN#1]
## Summary

On SM90, `na1d(..., backend="hopper-fna")` in FP16 returns `NaN` in `dQ` when
two things hold at once:

1. the token extent is **not** a whole number of backward KV tiles, so the last
   KV tile is partial; and
2. the products `O * dO` leave FP16's range (largest finite value 65504).

`dK` and `dV` stay finite, and the forward output is finite. The
`cutlass-fna` backend is unaffected under the identical inputs.

The affected positions are the last `kernel_size // 2` queries of `dQ`, for the
`kernel_size` values where the problem appears at all (3 and 7 in the sweep
below); `kernel_size = 15` showed none.

This is a second, separate site from the `dO * O` reduction overflow: that one
is in `FmhaKernelBwdSumOdO` and is fixed by converting both operands to FP32
before the product (`csrc/include/natten/cuda/reduction/fmha_kernel_bwd_sum_OdO.hpp`
and the Hopper and Blackwell copies of the same header). With that fix in place
the sweep below still reproduces, so the remaining overflow is an intermediate
inside the Hopper backward itself.

## Environment

| item | value |
| --- | --- |
| GPU | SM90 (`torch.cuda.get_device_capability() == (9, 0)`) |
| torch | 2.11.0+cu128 |
| NATTEN | `0.21.7+fork.3` and a build of `fix-compute-delta-precision` on top of it |
| CUDA | 12.8 |
| dtype | `torch.float16` |
| backend | `hopper-fna` |

## Minimal reproduction

```python
import torch
from natten.functional import na1d

BATCH, HEADS, HEAD_DIM = 2, 2, 32
EXTENT, KERNEL = 64, 7  # the backward KV tile here is 128, so 64 is partial

torch.manual_seed(42)
shape = (BATCH, EXTENT, HEADS, HEAD_DIM)
with torch.no_grad():
    q_ = torch.randn(shape, device="cuda", dtype=torch.float32) * 0.1
    k_ = torch.randn(shape, device="cuda", dtype=torch.float32) * 0.1
    v_ = torch.full(shape, 256.0, device="cuda", dtype=torch.float32)
    d_out_ = torch.full(shape, 512.0, device="cuda", dtype=torch.float32)

q = q_.half().requires_grad_(True)
k = k_.half().requires_grad_(True)
v = v_.half().requires_grad_(True)

out = na1d(q, k, v, kernel_size=(KERNEL,), backend="hopper-fna")
out.backward(d_out_.half())

nan_queries = (
    torch.isnan(q.grad.float())
    .any(dim=0).any(dim=-1).any(dim=-1)
    .nonzero().flatten().tolist()
)
print("out finite:", torch.isfinite(out.float()).all().item())
print("dQ NaN at queries:", nan_queries)
print("dK finite:", torch.isfinite(k.grad.float()).all().item())
print("dV finite:", torch.isfinite(v.grad.float()).all().item())
```

Output:

```
out finite: True
dQ NaN at queries: [61, 62, 63]
dK finite: True
dV finite: True
```

Every output element is 256 and every incident gradient is 512, so every
`O * dO` is 131072, about twice FP16's largest finite value.

## Sweep

`dQ` NaN query positions, `batch=2`, `heads=2`, `head_dim=32`, FP16,
`v = 256`, `dO = 512`. The backward KV tile for this configuration is 128
(`HOPPER_BACKWARD_CONFIGS[1][16][32][0] == ((64,), (128,))`).

| extent | kernel_size 3 | kernel_size 7 | kernel_size 15 |
| --- | --- | --- | --- |
| 32 | 31 | 29, 30, 31 | none |
| 56 | 55 | 53, 54, 55 | none |
| 63 | 62 | 60, 61, 62 | none |
| 64 | 63 | 61, 62, 63 | none |
| 65 | 64 | 62, 63, 64 | none |
| **128** | **none** | **none** | **none** |
| 130 | 129 | 127, 128, 129 | none |

Extent 128 is exactly one KV tile and is clean; every other extent in the
sweep leaves a partial last tile and is affected for `kernel_size` 3 and 7.

## What makes it disappear

Measured on the build described above, one run per row
(`probe_iv_disappear.py`), all with `batch=2`, `heads=2`, `head_dim=32`,
`kernel_size=(7,)`:

| change from the reproduction above | dQ finite | dQ NaN queries | dK, dV finite |
| --- | --- | --- | --- |
| none (extent 64, FP16, `hopper-fna`, `v = 256`, `dO = 512`) | **no** | 61, 62, 63 | yes |
| extent 128 (exactly one KV tile) | yes | none | yes |
| `v = 128`, `dO = 256` (product 3.3e4) | yes | none | yes |
| `backend="cutlass-fna"`, everything else identical | yes | none | yes |
| `backend="cutlass-fna"` in FP32 | yes | none | yes |

The forward output is finite in every row.

## Registered test

`tests/test_hopper_fna.py::HopperFNAPartialKVTileRangeTest::test_backward_over_a_partial_last_kv_tile`
asserts `dQ` is finite and is marked `pytest.mark.xfail(strict=True)`, so it
reports as XFAIL today and fails the suite the moment the underlying problem is
fixed.
